receiver/vmwarevcenterreceiver/scraper.go

```diff
@@ -0,0 +1,364 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vmwarevcenterreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vmwarevcenterreceiver"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vsan/types"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vmwarevcenterreceiver/internal/metadata"
+)
+
+// example 2022-03-10 14:15:00
+const timeFormat = "2006-01-02 15:04:05"
+
+var _ component.Receiver = (*vcenterMetricScraper)(nil)
+
+type vcenterMetricScraper struct {
+	client      *vmwareVcenterClient
+	mb          *metadata.MetricsBuilder
+	logger      *zap.Logger
+	vsanEnabled bool
+}
+
+func newVmwareVcenterScraper(
+	logger *zap.Logger,
+	config *Config,
+) *vcenterMetricScraper {
+	client := newVmwarevcenterClient(config)
+	return &vcenterMetricScraper{
+		client:      client,
+		logger:      logger,
+		mb:          metadata.NewMetricsBuilder(config.MetricsConfig.Metrics),
+		vsanEnabled: true,
+	}
+}
+
+func (v *vcenterMetricScraper) Start(ctx context.Context, _ component.Host) error {
+	return v.client.Connect(ctx)
+}
+
+func (v *vcenterMetricScraper) Shutdown(ctx context.Context) error {
+	return v.client.Disconnect(ctx)
+}
+
+func (v *vcenterMetricScraper) scrape(ctx context.Context) (pdata.Metrics, error) {
+	if v.client == nil {
+		return pdata.NewMetrics(), errors.New("failed to connect to http client")
+	}
```

Comment:
The `Connect` method is very complex so this is general guidance:

We need to be tolerant of potentially recoverable failures. Returning an error here will stop the collector process, which is not the desired behavior unless there is something unrecoverable. 

It's ok to log an error in `Start` if the failure is recoverable, but `scrape` should attempt a connection if there's a chance it will work. (eg. temporary network issue is resolved)